### ISSUE TITLE
Event stream: Really a golden source?

### ISSUE THREAD

**Issue by k0001** [cdodev/design#1]
Regarding the following:

> The main difference of a functional app is that we don't mutate state directly. In fact current state is derived entirely from events that have been observed to happen, so **we should be able to recover any state by replaying the events we have seen**. The stream of events that has occurred is the golden source of data in the system and can be folded down into any other data source to express state.

I am hesitant about the part that I highlighted. In theory, that is true. In practice, recovering events that were saved sometime in the past is incredibly hard, as we already know from our experience at Fynder. There are some workarounds such as keeping state snapshots, or attempting to have some kind of migrations for events themselves, but these are just ad-hoc kludges and they can't really be generalized in a reliable manner. 

I suggest we leave out the idea of “recovering any state by **replaying** the events we have seen”. We should have events, yes, but consider them “messages” as in a “message passing”, and use that to drive changes to the UI in real time (say, through a PubSub API over WebSockets), but those messages are never stored in a database, they just exist in memory and distributed over the network until they are not needed anymore.

**Comment by commandodev** [comment]
To be completely principled about this, we could regard changes in schema as events themselves. After all the event stream does represent the sum total of what we know about a piece of state at a point in time, regardless of whether we can de-serialize it in practice later on on! Even though we had the problem with changing schemas, I'm hesitant to throw the whole idea out :-)

If we break up the event so that we have one event type per module rather than the huge ADT we ended up with at fynder, perhaps adding constructors as our domain changes, rather than try to do some kind of migration would be a bit more tractable. What do you think?

I think I was a bit imprecise when I said "replaying". I think that we should keep all the events around, but in practice "replaying" might equate to folding with the last valid snapshot as the seed, rather than replaying the whole stream with some empty or starting state object.

Apache avro does seem to be able to handle some schema migration over time: https://avro.apache.org/docs/current/#compare

**Comment by k0001** [comment]
> perhaps adding constructors as our domain changes, rather than try to do some kind of migration would be a bit more tractable.

The problem is not in that, but in giving a sensible meaning to constructors that were used in the past.

> I'm hesitant to throw the whole idea out :-)

I'm not sure if the idea is that worthy. I mean, you still get the benefits of being able to fold the event stream as long as your application is running, it's just that you remove the possibility of folding everything since the beginning of time, instead you are only able to fold what's happening at the moment, starting from a “snapshot” that is basically the current state of your database when your program starts. 

Suppose somewhere in your website frontend you decide to create an FRP widget that displays 
the current number of friends. Here is some pseudocode using Reflex for this:

``` haskell
-- First an RPC call to get the current number of friends. This will trigger a lookup in a
-- relational database on the server side, say. “A lookup in the snapshot database” if you will.
n0 :: Int <-  getCurrentNumberOfFriendsForUserId myUserId

-- Second, we subscribe in a FRP manner to changes to this number. 
-- Changes will reach the client through some kind of PubSub API, 
-- driven by the "event source" on the server-side we've been discussing.
--
-- `Event` here is from `reflex`, and its name might be confusing given the other type of
-- “event”s we have been talking about, but you get the idea.
en :: Event t Int <- subscribeToChangesToNumberOfFriends myUserId

-- Third you pack all of that in an FRP dynamic value.
dn :: Dynamic t Int <- holdDyn n0 en

-- Now you are free to use `dn` to drive an FRP UI, with `dn` itself driven by the 
-- event source we've been discussing.
```

Seeing that this suffices, my question is: Why would we need to maintain an event stream that goes back till the very first event? Or even one that tries to model what a “snapshot” is? What do we gain from this? Aren't we better out just giving complete freedom over the choice of `a` in `foldl :: (a -> b -> a) -> a -> [b] -> a` (`foldl` is simplified, of course, in practice we would be generating descriptions for changes as we go, but you get the idea)?

**Comment by commandodev** [comment]
Here's a motivating example:

At a hedge fund I worked at we had a very event driven model. A schedule would kick off a number of actions in sequence:
- Fetch market price
- Run algo (which by the way was also just a fancy scanl)
- Scale signal by sizes of accounts to generate desired trades
- Electronically execute trades (also event driven - but a separate system)
- Allocate filled trades
- Archive completed trade record

We modeled this as passing a token (called a ticket) around that had information added to it at each stage (update statements pretty much) until we ended up with a full ticket at the end. This was the equivalent of throwing all the events away after folding over them.

This worked fine, until we wanted to to do some analysis on how long it took to get a ticket between two point (or fold the events a different way). Because we didn't keep all the actual events that happened, we were reduced to grepping log files pretty much. If we had architected our system as a fold over events, then we would have had to simply write a different fold to get the answers we needed.

This is why I consider the event stream the golden source: because it contains information that it valuable, even if you only realise that value at a later date. Also keeping all the events around doesn't preclude optimizations like folding forward from the last snapshot.

**Comment by commandodev** [comment]
This is an interesting take on essentially the same problem:

http://www.confluent.io/blog/introducing-kafka-streams-stream-processing-made-simple

**Comment by k0001** [comment]
Ben, I truly appreciate the benefits of event-source-like solutions, they are great and we should use that. The problem is not in event streams per se, the problem is in storing them: I don't know of any reliable way of maintaining past events in the event stream meaningful over time, considering some of them stop making sense as soon as your model changes. We've been through this at Fynder and eventually we decided to abandon the guarantee that one could “replay” the event stream, and instead we just kept the event stream stored just as a log intended for human consumption only, not for computers. 

On the other hand, using events to drive an FRP-like system is a fantastic idea, it is just storing them and being able to interpret them meaningfully in the future what poses a problem.

In other words: I have no idea how to solve this in a reliable, efficient (from a running time, but most importantly from a development time point of view), and not boilerplate-y way. In fact, depending on how your model changes over time, there might not be an obvious solution at all to how to deal with that change in while interpreting the event stream.

**Comment by commandodev** [comment]
Have you got a concrete example of where that was an insurmountable problem at fynder? I think it might have been after my time when you finally hit a case that meant you could no longer interpret the event stream.

Maybe we could enumerate all the ways in which a model can change over time (add fields, remove fields, rename, etc...) and the same for events. That way could maintain a matrix of situations we can handle. Perhaps a good most 80% solution would be to be able to say "we can handle `A` and `B` scenario, but `C` is a phase change and you then need to do this"? (`A` might be something like add a nullable field, `C` might be something like our whole domain model has changed)

On reflection, I accept that events don't have to be interpretable back to the beginning of time - if nothing else there will come a point in a large system where it's computationally infeasible. I also accept that the value of the data in an event decays the older it is. However I don't think an event needs to be de-serialisable to the same thing it's always been known as to have value. Consider the analytics dashboard I put together at fynder: it was basically a join of the event stream to the current state. Just SQL. It was the granularity of the data that made that possible.

Just for the sake of completeness: what do you think it wrong with creating a new event type or constructor when you hit a `C` like phase shift? Say you have

``` haskell
data MyEvent =
    CreateState State
  | ChangeFieldA StateId NewA
```

and you go to

``` haskell
data MyEvent
    CreateStateOld State
  | CreatState NewStateType
```

Your fold function would just need an extra case.

Worst comes to worst you just duplicate the entire model and state and the old one represents how we understood the world up to a phase shift and the new one is the new order?

I know I'm being quite hand wavy and not very concrete here, but I like to go top down when sketching out initial designs like this and cling to generalities for as long as possible (sometime too long) 

:smiley:

**Comment by k0001** [comment]
Ben, I'm in my phone now, so I can't type much, but one other thing to keep
in mind is: even if at times you can maintain backward compatibility with
past events, the amount of boilerplate necessary to interpret those past
representations is huge, and needs to be maintained and tested manually. It
slows downs development time significantly, and I'm sure you'll remember
some of that from Fynder. And yes, we abandoned this support shortly after
you left (and it was a productivity boost). I'll try to come up with some
examples later, but meanwhile you can think of a model being split into
many others over time, or foreign keys moving around, or just the meaning
of a model or field changing.

**Comment by commandodev** [comment]
WIP - I'm going to add to this or move it to a separate document.

I've been doing some reading up and I thought it might be useful to lay out our options as I see them. First some assumption (which may be invalid):

### Assumptions

##### We want events stored as JSON in a postgres table

It was very useful to be able to dig around and find patterns using SQL. However - perhaps instead of primary storage, we should treat the events table as a projection like any other and use a binary format for actual golden source storage.

### Prior art

For the sake of completeness here are the existing approaches I've found (some of which I know we looked at and discounted)
- Safecopy: https://hackage.haskell.org/package/safecopy-0.9.0.1/docs/Data-SafeCopy.html
- Protobuf: Supports field renaming and adding `Maybe` fields
- Avro: https://avro.apache.org/docs/current/ Seems to support a wider range of migrations that PB. Poor Haskell support at the moment. (Another article: http://blog.cloudera.com/blog/2011/05/three-reasons-why-apache-avro-data-serialization-is-a-good-choice-for-openrtb/)
- SchemaVer: http://snowplowanalytics.com/blog/2014/05/13/introducing-schemaver-for-semantic-versioning-of-schemas/ Not a solution on its own, but may be part of the solution
- "Up casting": some kind of type level read migration: https://abdullin.com/post/event-sourcing-versioning/ (.NET example)
  - Whatever groundhog is doing here: https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/groundhog (only skimmed this so far)
- Some kind of type level polymorphism like @tomjaguarpaw  uses for Opaleye. (See @Gabriel439's comment here http://stackoverflow.com/questions/18266137/extensible-serialization-in-haskell)
- File visioning: http://nhlab.blogspot.co.uk/2008/12/data-migration-with-happs-data.html
- Move versioning to the infrastructure level with something like iglu http://snowplowanalytics.com/documentation/concepts/iglu/

The snowplow guys have been done a lot of interesting stuff with events for a while (more from the downstream event consumption/data science angle admittedly), we might have more to learn from them.

**Comment by Gabriella439** [comment]
You might also want to check out my incomplete [`event-source` repo](https://github.com/Gabriel439/Haskell-Event-Source-Library/blob/master/src/Event/Source.hs#L469) which has a sketch of one way to approach the problem

**Comment by commandodev** [comment]
Thanks @Gabriel439, much appreciated! What change have you made to foldl here that requires duplicating it?

**Comment by Gabriella439** [comment]
The main change was to add a `Serialize` constraint to the internal `Monoid`, which I might eventually upstream into `foldl` at some point, but I was hoping to wait until one of the more efficient serialization libraries would come out first.
